sort node roles

paivagustavo opened the pull request:
Makes node roles to be sorted so it doesn't change randomly when being displayed.


fixes #344

derailed commented:
@paivagustavo Thank you for taking this on! 

A few notes...

* Not taking the sets.String as a parameter and returning in will cause an unwanted allocation. 
* If the role is empty, we need to return a []string{MissingValue} and not the empty slice so we keep parity with the current implementation not to mention kubectl.
* In the test we should probably have a use case for the alternate label "kubernetes.io/role" to make sure we're picking it up correctly. 

Thank you Gustavo!

paivagustavo commented:
> Not taking the sets.String as a parameter and returning in will cause an unwanted allocation.

I'm not 100% sure if this is the better way to go. AFAIK, maps are always heap-allocated and sets.String is an alias to map. But I have changed to accomodate to your suggestion.

> If the role is empty, we need to return a []string{MissingValue} and not the empty slice so we keep parity with the current implementation not to mention kubectl.

Totally missed that. Added a new test case to ensure this in the future :)

> In the test we should probably have a use case for the alternate label "kubernetes.io/role" to make sure we're picking it up correctly.

Agreed, I've added more test cases.

derailed commented:
@paivagustavo Thank you so much for making these changes! When it comes to software engineering I firmly believe `They're no gods!`. So please take the following comment lightly as it is solely FYI and not a measure of competency. Always be weary of statements like `always`.  Maps don't necessarily yield an allocation if the compiler can assert the map can fit on the ~2k heap. 

```go
func f1() map[string]struct{} {
	m := make(map[string]struct{})
	m["fred"] = struct{}{}
	for k := range m {
		_ = k
	}
	return m
}

func f(m map[string]struct{}) {
	m["fred"] = struct{}{}
	for k := range m {
		_ = k
	}
}
```

yields

```text
BenchmarkF-8            24608900                47.2 ns/op             0 B/op          0 allocs/op
BenchmarkF1-8            7666917               145 ns/op             192 B/op          2 allocs/op
```

I am being a stickler here solely because we're sitting on a 2 secs refresh and believe managing GC pressure is important in this case. 

Another small comment on the tests. I think we should use a map vs a slice for the use cases as it will be easier for us to track down which use case failed in case we need to revisit in the future and use cases order here does not matter ;)

I hope these make sense! Thank you Gustavo!

paivagustavo commented:
Wow! I totally got it wrong. I thought it was always heap because of some old issue on Go Github repository, that must be true for some old version. Thanks for being patient and giving an example, I really appreciate Fernand!

If you want we can change the tests in a follow up PR.

derailed commented:
My pleasure! Thank you @paivagustavo!!
I did make the test slice->map changes and will push on the next drop.
